### ISSUE TITLE
Classes and resources with a single letter component in their package name cannot be loaded from jar files with /./ or /../ in their path

### ISSUE THREAD

**Issue by fangjinuo** [spring-projects/spring-boot#17341]
our project load mysql 8 class fail:

![loadmysqlfail](https://user-images.githubusercontent.com/8428331/60338822-93e93d00-99d9-11e9-8723-e758b46efc67.png)

package config:

```xml
<plugin>
      <groupId>org.spring.framework.boot</groupId>
      <artifactId>spring-boot-maven-plugin</artifactId>
      <executions>
             <execution>
                    <goals>
                         <goal>repackage</goal>
                    </goals>
             </execution>
      </executions>
      <configuration>
            <mainClass>our main class</mainClass>
            <layout>ZIP</layout>
            <excludes>
                  <exclude>
                         <groupId>mysql</groupId>
                         <artifactId>mysql-connector-java</artifactId>
                  </exclude>
            </excludes>
      </configuration>
</plugin>
```
project dist directory:
```
/--bin/
 --config/
 --lib/
     --drivers/
          -- mysql-connector-java-8.0.15.jar
     --bootstrap.jar
 --scripts/
```

by debug, I found when load 'com/mysql/cj/protocol/a/result/NativeResultset.class'， 
the spring boot custom  URLStremHandler  (***org/springframework/boot/loader/jar/Handler.java***  
[Handler.java](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/Handler.java)  ) will parse 'com/mysql/cj/protocol/a/result/NativeResultset.class' as 'com/mysql/cj/protocol/result/NativeResultset.class'  。

In other words, every single letter directory will be parse as empty string "";

**Comment by wilkinsona** [comment]
Thanks for the report. Which version of Spring Boot are you using?

**Comment by fangjinuo** [comment]
we can reproduce this bug in 2.1.3.RELEASE  ~ 2.1.6.RELEASE versions

**Comment by fangjinuo** [comment]
only windows has the bug

**Comment by wilkinsona** [comment]
I've just noticed that your using a zip layout and a custom dist directory structure. To ensure that we are looking at the same situation as you, please provide a minimal sample that reproduces the problem, either as a zip attached to this issue or in a separate repository on GitHub.

**Comment by springborland** [comment]
Fix this issue at https://github.com/springborland/spring-boot/commit/b67c7148fa6695956c8835955ebd9e8287088c27.

**Comment by wilkinsona** [comment]
Thanks, @springborland. Before we consider a fix, I'd like to make sure the problem is fully understood. To do that, I'd really like to see a sample from @fangjinuo that reproduces the problem.

**Comment by fangjinuo** [comment]
Based on debug,  I simplify the reproduction of the issue:

step1 : create an empty maven project, with spring-boot-loader dependency:
```xml
    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-loader</artifactId>
            <version>2.1.6.RELEASE</version>
        </dependency>
    </dependencies>
```
step2: create a test case:
```java
import org.springframework.boot.loader.LaunchedURLClassLoader;
import org.springframework.boot.loader.jar.JarFile;

import java.io.File;
import java.net.URL;

public class SpringBootClassLoaderBugReproduce extends LaunchedURLClassLoader {
    public SpringBootClassLoaderBugReproduce(URL[] urls, ClassLoader parent) {
        super(urls, parent);
    }

    public static void main(String[] args) throws Exception {
        // mysql jdbc driver should not in the classpath of a project
        // To reproduce the issue, the path should has xxx/../yyy
        File file = new File("E:/springbootbugreproduce/drivers/../drivers/mysql-connector-java-8.0.15.jar");
        System.out.println(file.exists());
        // spring boot's JarFile class
        JarFile jarFile = new JarFile(file);
        SpringBootClassLoaderBugReproduce reproduceCL = new SpringBootClassLoaderBugReproduce(new URL[]{jarFile.getUrl()}, SpringBootClassLoaderBugReproduce.class.getClassLoader());
        Class clazz = reproduceCL.findClass("com.mysql.cj.protocol.a.result.NativeResultset");
        System.out.println(clazz);
    }

}

```

step 3:  put a mysql 8 jdbc driver in the specified directory: <br/>
![mysql_driver_loaction](https://user-images.githubusercontent.com/8428331/60378655-1c0e2780-9a58-11e9-9a54-41ded412cda2.png)

step 4: execute the test case, the CLI result is:
```
true
Exception in thread "main" java.lang.ClassNotFoundException: com.mysql.cj.protocol.a.result.NativeResultset
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at SpringBootClassLoaderBugReproduce.main(SpringBootClassLoaderBugReproduce.java:20)

```

**Comment by stilljavain2019** [comment]
You guys are certainly master of [Spring-boot](https://springtutor.com/spring-boot-configuration-with-jasypt/).
